crates/bitcoin/src/formatter.rs

```diff
@@ -6,234 +6,364 @@ use crate::{merkle::MerkleProof, script::*, types::*, Error, GetCompact};
+// Generate `TryFormat` implementation for the basic integers (signed and unsgined) types
+make_try_format_int!(u8);
+make_try_format_int!(u16);
+make_try_format_int!(u32);
+make_try_format_int!(u64);
+make_try_format_int!(i8);
+make_try_format_int!(i16);
+make_try_format_int!(i32);
+make_try_format_int!(i64);
 
-impl Formattable<()> for bool {
-    fn format(&self) -> Vec<u8> {
-        (*self as u8).format()
+impl TryFormat<()> for bool {
+    fn format_size(&self, _: ()) -> u32 {
+        sp_std::mem::size_of::<u8>() as u32
+    }
+
+    fn try_format(&self, length_bound: &mut u32) -> Result<Vec<u8>, Error> {
+        (*self as u8).try_format(length_bound)
     }
 }
 
-impl Formattable<()> for H256Le {
-    fn format(&self) -> Vec<u8> {
-        Vec::from(&self.to_bytes_le()[..])
+impl TryFormat<()> for H256Le {
+    fn format_size(&self, _: ()) -> u32 {
+        Self::len_bytes() as u32
+    }
+
+    fn try_format(&self, length_bound: &mut u32) -> Result<Vec<u8>, Error> {
+        length_bound.checked_reduce(self.format_size(()))?;
+        Ok(Vec::from(&self.to_bytes_le()[..]))
     }
 }
 
-impl Formattable for CompactUint {
-    fn format(&self) -> Vec<u8> {
+impl TryFormat for CompactUint {
+    fn format_size(&self, _: ()) -> u32 {
+        match self.value {
+            0..=0xFC => 1,
+            0xFD..=0xFFFF => 3,
+            0x10000..=0xFFFFFFFF => 5,
+            _ => 9,
+        }
+    }
+
+    fn try_format(&self, length_bound: &mut u32) -> Result<Vec<u8>, Error> {
         let mut formatter = Formatter::new();
         if self.value < 0xfd {
-            formatter.format(self.value as u8);
+            formatter.try_format(self.value as u8, length_bound)?;
         } else if self.value < u16::max_value() as u64 {
-            formatter.format(0xfd_u8);
-            formatter.format(self.value as u16);
+            formatter.try_format(0xfd_u8, length_bound)?;
+            formatter.try_format(self.value as u16, length_bound)?;
         } else if self.value < u32::max_value() as u64 {
-            formatter.format(0xfe_u8);
-            formatter.format(self.value as u32);
+            formatter.try_format(0xfe_u8, length_bound)?;
+            formatter.try_format(self.value as u32, length_bound)?;
         } else {
-            formatter.format(0xff_u8);
-            formatter.format(self.value);
+            formatter.try_format(0xff_u8, length_bound)?;
+            formatter.try_format(self.value, length_bound)?;
         }
-        formatter.result()
+        Ok(formatter.result())
     }
 }
 
-impl<T, U> Formattable<U> for Vec<T>
+impl<T, U> TryFormat<U> for Vec<T>
 where
-    for<'a> &'a T: Formattable<U>,
+    for<'a> &'a T: TryFormat<U>,
     U: Default + Copy,
 {
-    fn format(&self) -> Vec<u8> {
-        self.format_with(Default::default())
+    fn format_size(&self, options: U) -> u32 {
+        let mut size = CompactUint::from_usize(self.len()).format_size(());
+        for value in self.iter() {
```

Comment:
To be pedantic, this technically is unbounded execution - you iterate over all elements of the vector without checking that the user paid weight for it. Unfortunately, fixing this requires another significant rewrite.. I can think of 2 ways to fix it:

(1) pass the length bound to `format_size` as well, or
(2) remove `format_size` as a trait method. I don't think this method it really necessary anyway - you can just call format_size, and have the call fail when it uses up its budget.

My preference would be 2

Reply:
This is technically unused by the runtime, but I have refactored the formatting code to closely resemble that of `rust-bitcoin`. Using this approach we can pass a `BoundedWriter` to fail as soon as the budget has been consumed.